app/src/main/java/com/toolslab/noisepercolator/db/DataProvider.kt

```diff
@@ -19,36 +22,88 @@ class DataProvider(private val context: Context = NoisePercolator.applicationCon
 
         @VisibleForTesting
         const val MESSAGES_KEY = "MESSAGES_KEY"
+
+        const val SHARED_PREFERENCES_DEPRECATED_MESSAGE = "Now using Realm instead of SharedPreferences."
     }
 
     @CheckResult
-    fun getMessages(): List<Message> {
+    fun getMessages(): Observable<List<Message>> {
+        migrateOldDatabaseIfNeeded()
+        val realm = realmWrapper.getDefaultInstance()
+        return realm.where(Message::class.java).findAllAsync()
+                .asFlowable()
+                .toObservable()
+                .map {
+                    realm.copyFromRealm(it)
+                }
+                .doOnDispose {
+                    // Close database when observable is not needed anymore
+                    realm.close()
+                }
+                .doOnError {
+                    // TODO ASK Close database?
+                    Timber.e(it, "Error in DataProvider.getMessages()")
+                }
+    }
+
+    fun saveMessages(messages: List<Message>) {
+        messages.forEach { saveMessage(it) }
+    }
+
+    fun saveMessage(message: Message) {
+        // Get a Realm instance for this thread
+        val realm = realmWrapper.getDefaultInstance()
+        // Persist your data in a transaction
+        realm.executeTransaction {
+            // Using executeTransaction with a lambda reduces code size
+            // and makes it impossible to forget to commit the transaction.
+            it.copyToRealm(message)
+        }
+        // Close database after saving
+        realm.close()
+    }
+
+    @Deprecated("$SHARED_PREFERENCES_DEPRECATED_MESSAGE Use DataProvider.getMessages().",
+            ReplaceWith("getMessages()", "com.toolslab.noisepercolator.db.DataProvider.getMessages"))
+    fun getMessagesFromSharedPreferences(): List<Message> {
         val messagesAsStringSet = getMessagesStringSet()
         return messagesAsStringSet.map { messageConverter.convert(it) }
     }
 
+    @Deprecated(SHARED_PREFERENCES_DEPRECATED_MESSAGE)
     @VisibleForTesting
-    @CheckResult
     fun getMessagesStringSet(): MutableSet<String> {
         return getPreferences().getStringSet(MESSAGES_KEY, mutableSetOf())
     }
 
-    fun saveMessage(message: Message) {
+    @Deprecated("$SHARED_PREFERENCES_DEPRECATED_MESSAGE Use DataProvider.saveMessage(message).",
+            ReplaceWith("saveMessage(message)", "com.toolslab.noisepercolator.db.DataProvider.saveMessage"))
+    fun saveMessageToSharedPreferences(message: Message) {
         val messages = getMessagesStringSet().toMutableSet()
         messages.add(messageConverter.convert(message))
         getPreferences().edit {
             putStringSet(MESSAGES_KEY, messages)
         }
     }
 
-    fun clearPreferences() {
+    private fun migrateOldDatabaseIfNeeded() {
```

Comment:
That would make more sense, I was just playing with a way of migrating instead of just starting with an empty database. Since there are no live users this implementation will disappear very soon.
I was learning to use the nice @Deprecated annotation of Kotlin together with Android Studio's autocomplete instead of deleting the implementation directly.